core.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"fmt"
+	"sync"
+)
+
+const _defaultCollectionSize = 128
+
+// A coreRegistry is a collection of metrics. Uniqueness is enforced with two
+// checks, just like the vanilla Prometheus client:
+//
+// First, any two metrics with the same name must have the same label names
+// (both constant and variable).
```

Comment:
The Prometheus client's logic on this is actually not very well thought-out - there's no reason to include the help text in the hashing logic, since these rules already prohibit multiple metrics with the same name but different help text. This is covered in the unit tests for the registry.

Reply:
cool, do we plan to implement prometheus compatibility without using their registry?

Reply:
Also, is it possible to create a counter and a gauge with the same labels and name?

Reply:
We're getting Prom compatibility without using their registry, but we *are* using their HTTP handler (which is separate from the registry). The registry doesn't actually do that much, but the HTTP handler does quite a bit that I don't want to replicate.

It's not possible to create a counter and a gauge that share the same name and labels.